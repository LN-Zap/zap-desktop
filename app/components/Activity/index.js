import Activity from './Activity'

export default Activity
