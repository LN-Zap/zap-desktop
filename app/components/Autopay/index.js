import Autopay from './Autopay'

export default Autopay
export AutopayActions from './AutopayActions'
export AutopayCardView from './AutopayCardView'
export AutopayHeader from './AutopayHeader'
export AutopayHeading from './AutopayHeading'
export AutopayList from './AutopayList'
export AutopaySearch from './AutopaySearch'
export AutopaySearchNoResults from './AutopaySearchNoResults'
