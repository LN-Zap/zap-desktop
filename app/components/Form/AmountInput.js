import React from 'react'
import PropTypes from 'prop-types'
import CurrencyIcon from 'components/CurrencyIcon'
import styles from './RequestForm.scss'

class AmountInput extends React.Component {

  constructor(props) {
    super(props)

    this.handleChange = this.handleChange.bind(this)
    this.handleKeyDown = this.handleKeyDown.bind(this)
  }

  parseNumber(_value) {
    let value = _value.replace(/[^0-9,.]/g, '')
    let integer = null
    let fractional = null

    if (value == '') { value = '0' }
    if (value*1.0 < 0) { value = '0.0' }

    // simplify this, make flat if possible
    // extract the parsing bit?
    if (value.match(/^[0-9]*[.,][0-9]*$/)) {
      [integer, fractional] = value.toString().split(/[.,]/)
      if (!fractional || fractional == '') {
        fractional = ''
      }
    } else {
      integer = value
    }

    // limit fractional precision to 8 places
    if (fractional && fractional.length > 8) {
      fractional = fractional.substring(0, 8)
    }

    return [integer, fractional]
  }

  formatValue(integer, fractional) {
    let value;
    if (fractional && fractional.length > 0) {
      value = `${integer}.${fractional}`
    } else { 
      if (fractional === '') {
        value = `${integer}.`
      } else {
        value = `${integer}`
      }
    }
    return value
  }

  handleChange(e) {
    const _value = e.target.value
    const [integer, fractional] = this.parseNumber(_value)
    let value = this.formatValue(integer, fractional)
    
    // reset cursor position here?
    if (value != _value) {
      this.resetCursorPosition = e.target.selectionEnd
    }

    this.props.onChange(value)
  }

  componentDidUpdate() {
    if (this.resetCursorPosition) {
      this.refs.input.setSelectionRange(this.resetCursorPosition, this.resetCursorPosition)
      this.resetCursorPosition = null
    }
  }

  shiftValue(value, delta) {
    let [integer, fractional] = this.parseNumber(value)

    if (fractional) {
      delta = delta / (10 ** fractional.length)
      value = (parseFloat(value) + delta).toFixed(fractional.length)
    } else {
      value = (integer*1 + delta).toString()
    }

    if (value < 0) { 
      if (fractional && fractional.length > 0) {
        value = (0.0).toFixed(fractional.length)
      } else {
        value = '0'
      }
    }

    return value
  }

  handleKeyDown(e) {
    let value = e.target.value

    if (e.key === 'ArrowUp') {
      e.preventDefault()
      value = this.shiftValue(value, 1)
      this.props.onChange(value)
      return
    }

    if (e.key === 'ArrowDown') {
      e.preventDefault()
      value = this.shiftValue(value, -1)
      this.props.onChange(value)
      return
    }

    // do not allow multiple commas or dots
    if ((e.key === '.') || (e.key === ',')) {
      if (value.search(/[.,]/) >= 0) {
        e.preventDefault()
      }
      return
    }

    if ((e.key.length == 1) && !e.key.match(/^[0-9.,]$/)) {
      e.preventDefault()
      return
    }
  }

  render() {
    const {
      amount,
      currency,
      crypto,
    } = this.props

    let fontSize = 170 - (amount.length ** 2)
    if (fontSize < 30) {
      fontSize = 30
    }

    // for width to work, we need the font to be monospace or more margin for error
    const inputStyle = { 
      width: fontSize/1.5 * amount.length,
      fontSize: fontSize,
    }

    return (
      <section className={styles.amountContainer}>
        <label htmlFor='amount'>
          <CurrencyIcon currency={currency} crypto={crypto} />
        </label>
        <input
          ref='input'
          type='text'
          size=''
          style={inputStyle}
          value={amount}
          onKeyDown={this.handleKeyDown}
          onChange={this.handleChange}
          id='amount'
        />
      </section>
    )
  }
}

AmountInput.propTypes = {
  currency: PropTypes.string.isRequired,
  crypto: PropTypes.string.isRequired,
  onChange:PropTypes.func.isRequired,
  amount: PropTypes.string.isRequired,
}

export default AmountInput
