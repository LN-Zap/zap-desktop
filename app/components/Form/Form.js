import React from 'react'
import PropTypes from 'prop-types'

import Isvg from 'react-inlinesvg'
import x from 'icons/x.svg'

import Pay from './Pay'
import Request from './Request'

import styles from './Form.scss'

const FORM_TYPES = {
  PAY_FORM: Pay,
  REQUEST_FORM: Request
}

const Form = ({ formType, formProps, closeForm }) => {
  if (!formType) {
    return null
  }

  const FormComponent = FORM_TYPES[formType]
  return (
    <div className={styles.container}>
      <div className={styles.closeContainer}>
        <span onClick={closeForm}>
          <Isvg src={x} />
        </span>
      </div>
      <FormComponent {...formProps} />
    </div>
  )
}

Form.propTypes = {
  formType: PropTypes.string,
  formProps: PropTypes.object.isRequired,
  closeForm: PropTypes.func.isRequired
}

export default Form
