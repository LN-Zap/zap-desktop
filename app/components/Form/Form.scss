@import '../../variables.scss';

.outtercontainer {
  position: absolute;
  top: 0;
  bottom: 0;
  width: 100%;
  height: 100vh;
  background: $white;
  z-index: 0;
  opacity: 0;
  transition: all 0.5s;

  &.open {
    opacity: 1;
    z-index: 20;
  }
}

.innercontainer {
  position: relative;
  height: 100vh;
  margin: 5%;
}

.esc {
  position: absolute;
  top: 0;
  right: 0;
  color: $darkestgrey;
  cursor: pointer;
  padding: 20px;
  border-radius: 50%;

  &:hover {
    color: $bluegrey;
    background: $darkgrey;
  }

  &:active {
    color: $white;
    background: $main;
  }

  svg {
    width: 32px;
    height: 32px;
  }
}

.content {
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  height: 75vh;
  justify-content: center;
  align-items: center;
}
