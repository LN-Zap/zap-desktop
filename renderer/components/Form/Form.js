import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import system from '@rebass/components'
import { styles } from 'styled-system'
import { Form as InformedForm } from 'informed'

// Create an html input element that accepts all style props from styled-system.
const FormInner = system(
  {
    extend: InformedForm,
  },
  ...Object.keys(styles)
)

const Form = ({ asyncValidators, onSubmit, onSubmitFailure, getApi, ...rest }) => {
  const formApi = useRef(null)

  const onBeforeSubmit = async values => {
    // pause submit if we have async validators set
    if (asyncValidators && asyncValidators.length) {
      await Promise.all(asyncValidators.map(v => v()))
    }

    const api = formApi.current
    if (api) {
      if (api.getState().invalid) {
        onSubmitFailure && onSubmitFailure()
      } else {
        onSubmit && onSubmit(values)
      }
    }
  }

  const getApiWrapper = api => {
    formApi.current = api
    getApi && getApi(api)
  }

  return (
    <FormInner
      {...rest}
      getApi={getApiWrapper}
      onSubmit={onBeforeSubmit}
      onSubmitFailure={onSubmitFailure}
    />
  )
}

Form.propTypes = {
  asyncValidators: PropTypes.array,
  getApi: PropTypes.func,
  onSubmit: PropTypes.func,
  onSubmitFailure: PropTypes.func,
}

export default Form
