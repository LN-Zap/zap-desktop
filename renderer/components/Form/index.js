export { BasicInput } from './Input'
export { BasicRange } from './Range'
export { BasicTextArea } from './TextArea'
export { BasicCheckbox } from './Checkbox'
export { BasicSelect } from './Select'
export Checkbox from './Checkbox'
export CryptoAmountInput from './CryptoAmountInput'
export CurrencyFieldGroup from './CurrencyFieldGroup'
export FiatAmountInput from './FiatAmountInput'
export Form from './Form'
export Input from './Input'
export InputLabel from './InputLabel'
export Label from './Label'
export LightningInvoiceInput from './LightningInvoiceInput'
export LndConnectionStringInput from './LndConnectionStringInput'
export LndConnectionStringEditor from './LndConnectionStringEditor'
export NodePubkeyInput from './NodePubkeyInput'
export IntegerInput from './IntegerInput'
export OpenDialogButton from './OpenDialogButton'
export OpenDialogInput from './OpenDialogInput'
export Radio from './Radio'
export RadioCard from './RadioCard'
export RadioGroup from './RadioGroup'
export Range from './Range'
export RowLabel from './RowLabel'
export Select from './Select'
export TextArea from './TextArea'
export Toggle from './Toggle'
export TransactionFeeInput, { TransactionSpeedDesc } from './TransactionFeeInput'
