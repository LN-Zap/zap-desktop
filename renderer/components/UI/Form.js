import system from '@rebass/components'
import { styles } from 'styled-system'
import { Form as InformedForm } from 'informed'

// Create an html input element that accepts all style props from styled-system.
const Form = system(
  {
    extend: InformedForm,
  },
  ...Object.keys(styles)
)

export default Form
