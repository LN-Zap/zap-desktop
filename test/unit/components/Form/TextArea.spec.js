import React from 'react'

import { Form } from 'informed'

import { renderWithTheme } from '@zap/test/unit/__helpers__/renderWithTheme'
import { TextArea } from 'components/Form'

describe('component.UI.Input', () => {
  it('should render correctly', () => {
    const tree = renderWithTheme(
      <Form>
        <TextArea field="name" />
      </Form>
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
