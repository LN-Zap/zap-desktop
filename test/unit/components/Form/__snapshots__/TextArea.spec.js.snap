// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component.UI.Input should render correctly 1`] = `
.c0 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  padding: 16px;
  color: #ffffff;
  background-color: transparent;
  border: 1px solid;
  border-color: #959595;
  border-radius: 5px;
  font-family: Roboto,system-ui,sans-serif;
  font-size: 13px;
  font-weight: 300;
  width: 100%;
  opacity: inherit;
  outline: none;
  border-color: #959595;
}

.c2:not([readOnly]):not([disabled]):focus {
  border-color: #fd9800;
}

.c2::-webkit-search-decoration:hover,
.c2::-webkit-search-cancel-button:hover {
  cursor: pointer;
}

<form
  onKeyDown={[Function]}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <input
        className="c2"
        color="primaryText"
        disabled={false}
        fontFamily="sans"
        fontSize="m"
        fontWeight="light"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        readOnly={false}
        value=""
        width={1}
      />
    </div>
  </div>
</form>
`;
